qiskit/utils/deprecation.py

```diff
@@ -50,30 +51,74 @@ def wrapper(*args, **kwargs):
             warnings.warn(msg, category=category, stacklevel=stacklevel)
             return func(*args, **kwargs)
 
+        _DeprecationMetadata.set_func_deprecation(func, msg=msg, since="TODO")
         return wrapper
 
     return decorator
 
 
-def _rename_kwargs(func_name, kwargs, kwarg_map, category: Type[Warning] = DeprecationWarning):
+def _rename_kwargs(
+    func: Callable,
+    kwargs: Dict[str, Any],
+    kwarg_map: Dict[str, str],
+    category: Type[Warning] = DeprecationWarning,
+) -> None:
+    func_name = func.__name__
     for old_arg, new_arg in kwarg_map.items():
+        if new_arg is None:
+            msg = (
+                f"{func_name} keyword argument {old_arg} is deprecated and "
+                "will in the future be removed."
+            )
+        else:
+            msg = (
+                f"{func_name} keyword argument {old_arg} is deprecated and "
+                f"replaced with {new_arg}."
+            )
+        _DeprecationMetadata.set_args_deprecation(func, arg=old_arg, msg=msg, since="TODO")
         if old_arg in kwargs:
             if new_arg in kwargs:
                 raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
+            warnings.warn(msg, category=category, stacklevel=3)
+            if new_arg is not None:
+                kwargs[new_arg] = kwargs.pop(old_arg)
 
-            if new_arg is None:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    "will in future be removed.",
-                    category=category,
-                    stacklevel=3,
-                )
-            else:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    f"replaced with {new_arg}.",
-                    category=category,
-                    stacklevel=3,
-                )
 
-                kwargs[new_arg] = kwargs.pop(old_arg)
+@dataclass(frozen=True)
+class _DeprecationMetadataEntry:
+    msg: str
+    since: str
+
+
+@dataclass
+class _DeprecationMetadata:
+    """Used to store deprecation information on a function.
+
+    This is used by the Qiskit Sphinx Theme to render deprecations in documentation. Warning:
+    coordinate changes with the Sphinx Theme's extension.
```

Comment:
We can't do a fully synchronised release between the two packages (Terra CI depends on the sphinx theme, and I assume vice versa), so while it's good to note that this is only intended to be consumed by the Sphinx extension, we're going to have to manage changes to this structure the same way as any other inter-library API boundary. By that, I mean ensuring that each new feature version of Terra can still be understood by the most recent release version of the Sphinx theme, even if the theme won't be interpreting _all_ the information.  Since this is a private attribute, though, that won't ever be too difficult.

Reply:
Exactly, this is what I was trying to get at. I'll reword this for better clarity.